### ISSUE TITLE
Wrong error if a crate exports a macro with the same name as itself

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust#47361]
If a crate or module defines a macro with the same name as the crate or module, any type error suggests that you invoke a non-existent macro instead of what you meant to write.

```rust
mod syn {
    #[allow(unused_macros)]
    macro_rules! syn {
        () => {}
    }
}

fn main() {
    let _: syn::Foo;
}
```

```
error[E0573]: expected type, found macro `syn::Foo`
  --> src/main.rs:22:9
   |
22 |     let _: syn::Foo;
   |            ^^^^^^^^ did you mean `syn::Foo!(...)`?
```

If the macro name is different from the module name then the error is reasonable.

```
error[E0412]: cannot find type `Foo` in module `syn`
  --> src/main.rs:22:14
   |
22 |     let _: syn::Foo;
   |                 ^^^ not found in `syn`
```

This currently affects the Syn crate. @Arnavion

**Comment by dtolnay** [comment]
Mentioning @petrochenkov because the most relevant sounding change I could find is #38154.

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/38154 was written when macro paths with multiple segments weren't legal, so it made... assumptions.
The mistake is here:
https://github.com/rust-lang/rust/blob/73ac5d6a80f26c692f1e084b72d69637d7de2c8c/src/librustc_resolve/lib.rs#L2856.

**Comment by estebank** [comment]
Repro case and current output for case from #57588:

```rust
use std::env;

fn main() {
    env::vav_os("PATH");
}
```

```
error[E0423]: expected function, found macro `env::vav_os`
 --> src/main.rs:4:5
  |
4 |     env::vav_os("PATH");
  |     ^^^^^^^^^^^
help: a function with a similar name exists
  |
4 |     env::var_os("PATH");
  |          ^^^^^^
help: use `!` to invoke the macro
  |
4 |     env::vav_os!("PATH");
  |     ^^^^^^^^^^^^
```

Current output for the case in the original report:

```
error[E0573]: expected type, found macro `syn::Foo`
  --> src/main.rs:10:12
   |
10 |     let _: syn::Foo;
   |            ^^^^^^^^ help: use `!` to invoke the macro: `syn::Foo!`
```

**Comment by estebank** [comment]
Current output:

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=d64537ec9c6390f24998d9f5aeabb0f1

```
error[E0425]: cannot find function `vav_os` in module `env`
 --> src/main.rs:4:10
  |
4 |     env::vav_os("PATH");
  |          ^^^^^^ help: a function with a similar name exists: `var_os`
 --> /rustc/f3126500f25114ba4e0ac3e76694dd45a22de56d/library/std/src/env.rs:268:1
  |
  = note: similarly named function `var_os` defined here
```

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=a1153b2234ab364b7fc70b182128e153
```
error[E0412]: cannot find type `Foo` in module `syn`
 --> src/main.rs:9:17
  |
9 |     let _: syn::Foo;
  |                 ^^^ not found in `syn`
```
